### PR TITLE
Article UI testing

### PR DESCRIPTION
**Pull request by v1sar**

### CODE REVIEW

File: docs/content/basics/ui_testing.md [hunk 1]
```diff
@@ -1,4 +1,119 @@
 # UI testing
 
-To be done
+UI tests is a part of instrumentation tests.
```
Comment (NoNews) [inline docs/content/basics/ui_testing.md:3]: `UI testing is`
`UI tests are`

File: docs/content/basics/ui_testing.md [hunk 2]
```diff
@@ -1,4 +1,119 @@
...
+<br/>`UiAutomator` is a slower tool and can support tests only in black-box style, but it can test everything outside your application on your phone.
+<br/>Many cross-platforms tools for testing using `UiAutomator`.
+
+### Where to start
+
+Firstly, you should add dependency for Espresso library to your Android app module
+
+```kotlin
+dependencies {
+    androidTestImplementation('androidx.test.espresso:espresso-core:3.4.0')
```
Comment (NoNews) [inline docs/content/basics/ui_testing.md:26]: I'm not really sure that it's worth to have a documentation description here. We can just point to the android documentation

**Review by NoNews — CHANGES_REQUESTED** [review]
Being honest, don't really like that we post just regular documentation here 🤔 
Let's have more proper things.

From my perspective, this article:
- Instruments for UI testing
Short description for Espresso/UiAutomator, short points why they are bad, short examples, without how connect it to the gradle. (We have a documentation for that provided by google)

- Available tools for UI testing (review the existed libraries, like Barista, Kakao, Kaspresso, Layout Inspector + native viewer)

What do you think about it?


UPD

@sergio-sastre waiting for your opinion :)

File: docs/content/basics/ui_testing.md (hunk 2 above)
Comment (v1sar) [reply docs/content/basics/ui_testing.md]: allright, let's delete that

File: docs/content/basics/ui_testing.md (hunk 1 above)
Comment (v1sar) [reply docs/content/basics/ui_testing.md]: fixed

**Comment by v1sar** [comment]
So, answering to @NoNews comment:
* added info about ui testing in general
* Added more detailed info about `Espresso` and `UiAutomator` and their comparison
* Made more informative conclusion

My opinion is that this article should just show people the main purpose of Ui testing and short example with espresso is very good for beginners which will meet our cookbook. In conclusion i've added info where everybody can see detailed info about possible problems and which libraries may solve that. It will be quite strange to discard simple example on espresso (documentation-like) but promote documentation of external libraries :) 

let's decide what to do

**Comment by sergio-sastre** [comment]
> Being honest, don't really like that we post just regular documentation here 🤔 Let's have more proper things.
> 
> From my perspective, this article:
> 
> * Instruments for UI testing
>   Short description for Espresso/UiAutomator, short points why they are bad, short examples, without how connect it to the gradle. (We have a documentation for that provided by google)
> * Available tools for UI testing (review the existed libraries, like Barista, Kakao, Kaspresso, Layout Inspector + native viewer)
> 
> What do you think about it?
> 
> UPD
> 
> @sergio-sastre waiting for your opinion :)
I would mention the tools with bullet points and  link them to their respective sections  (not part of the MVP now I would say, so no links for now), or Github repos for now.

File: docs/content/basics/ui_testing.md [hunk 3]
```diff
@@ -1,4 +1,148 @@
...
+* Login to your application
+* Process of ordering 
+* Creating new chat 
+* And so on
+
+Simple Ui test can't catch the problem with wrong padding or wrong color of your view (only in cases where you specially check that) - because there is another type of tests used to catch such problems called `Screenshot testing` - you can read [Screenshot testing article](https://android-ui-testing.github.io/Cookbook/basics/screenshot_testing/) to get familiar with that.<br/>
+
+So, if once we decided that we need to check users scenarios - we should definitely start writing Ui tests and let's start with tools, that will help us to write this type of tests.
+
+## Main tool - Espresso and UiAutomator
```
Comment (sergio-sastre) [inline docs/content/basics/ui_testing.md:18]: Maybe worth mentioning that Jetpack Compose testing library for Composables?

"Different testing tools. Espresso and UIAutomator can still test a Compose Layout - searching by text, resource, etc. However, they don’t have access to Composables’ semantics and can’t fully test them. Therefore, it’s recommended to use the Jetpack Compose testing library as it can access semantics and fully test Composables on the screen."
Link:
https://dmytroshuba.com/blog/jetpack-compose-ui-testing/

File: docs/content/basics/ui_testing.md [hunk 4]
```diff
@@ -1,4 +1,148 @@
...
+<br/>But don't worry - you can use `UiAutomator` for that purpose, and it can be called directly from test, written in Espresso.
+<br/>`UiAutomator` is a slower tool and can support tests only in black-box style, but it can test everything outside your application on your phone.
+<br/>Many cross-platforms tools for testing using `UiAutomator`.
+
+These two tools are the main tools that everyone using for Ui testing on Android. <br/>
+Remember, it's possible to create tests that will use both frameworks. For example, you can find your views and interact with them by `Espresso` and in the same test call `UiAutomator` to perform something specific, like pressing volume button. <br/>
+But for usual testing you should prefer `Espresso` rather then `UiAutomator`. 
+
+### Espresso
+
```
Comment (sergio-sastre) [inline docs/content/basics/ui_testing.md:37]: Maybe also worth mentioning that Espresso as well as the Jetpack Compose Library support Robolectric, while UiAutomator doesn't.

That means, one can write Shared Tests written in Espresso or with Compose Library

File: docs/content/basics/ui_testing.md [hunk 5]
```diff
@@ -1,4 +1,148 @@
...
+Why Espresso is so cool? It is synchronized with the main thread of your app and performs action only when main thread is in idle state.
+<br/>It makes Espresso extremely reliable and stable (than any other existing tool) for Ui testing on Android.
+
+Keep in mind that everything that isn't related to your app (for example, permission dialogs, volume +/- buttons, home button and etc) - can't be accessed directly within Espresso.
+<br/>But don't worry - you can use `UiAutomator` for that purpose, and it can be called directly from test, written in Espresso.
+<br/>`UiAutomator` is a slower tool and can support tests only in black-box style, but it can test everything outside your application on your phone.
+<br/>Many cross-platforms tools for testing using `UiAutomator`.
+
+These two tools are the main tools that everyone using for Ui testing on Android. <br/>
+Remember, it's possible to create tests that will use both frameworks. For example, you can find your views and interact with them by `Espresso` and in the same test call `UiAutomator` to perform something specific, like pressing volume button. <br/>
```
Comment (sergio-sastre) [inline docs/content/basics/ui_testing.md:33]: Really like this :) Should be bold maybe? I would like to highlight it

File: docs/content/basics/ui_testing.md [hunk 6]
```diff
@@ -1,4 +1,148 @@
...
+        onView(withId(R.id.button))
+            .perform(click())
+
+        onView(withId(R.id.textview))
+            .check(matches(isDisplayed()))
+    }
+}
+```
+
+### What else?
```
Comment (sergio-sastre) [inline docs/content/basics/ui_testing.md:132]: I would rather name it more meaningfully:
Espresso Cheat Sheet

Comment (sergio-sastre) [reply docs/content/basics/ui_testing.md:132]: Moreover, there is the same for Compose, worth mentioning on its corresponding section
Compose Cheat sheet
https://developer.android.com/jetpack/compose/testing-cheatsheet

File: docs/content/basics/ui_testing.md [hunk 7]
```diff
@@ -1,4 +1,148 @@
...
+<br/>But there is a quite famous "cheat-sheet" for Espresso, I hope it will be helpful for everybody.
+
+![alt text](../images/ui_testing/ui_testing_cheat_sheet.jpeg "Cheat sheet")
+
+From that cheat-sheet you can understand which methods you can use for finding the `Views`, which methods for interactions with Views and which methods for checking the data.
+
+### Something like conclusion
+
+To be honest, there is a lot of edge-cases once you are trying to write your own tests. That's one of the reasons why that CookBook was created, so we are suggesting to read next articles to understand how to overcome many of common problems that you will face in Ui testing. <br/>
+Beside that fact, that `Espresso` and `UiAutomator` are the main tools to do Ui testing on Android, you may notice at [Companies experience article](https://android-ui-testing.github.io/Cookbook/adoption/companies_experience/) that almost nobody uses only these tools.<br/>
```
Comment (sergio-sastre) [inline docs/content/basics/ui_testing.md:144]: I like the point. I would also add something about "new way of testing" due to "Compose emerging".

Additionally, I would reword this though.
Something like "most companies do not use Espresso or UiAutomator *directly*, but through frameworks that wrap them under the hood:
1. Espresso 
   - Kakao
   - Barista
   - Kaspresso
2. UiAutomator
   - Kaspresso (Kautomator)
3. Compose Test Library
   - Kaspresso (since 1.4.0) 
   
Moreover you can write something about "Compose is emerging) and that Kaspresso has also added support for it in 1.4.0

**Review by NoNews — APPROVED** [review]
